src/main/java/it/gov/pagopa/payhub/activities/activity/treasury/TreasuryOpiIngestionActivityImpl.java

```diff
@@ -26,50 +27,82 @@
-    this.ingestionFlowFileRetrieverService = ingestionFlowFileRetrieverService;
-  }
 
+           List <String> iufList = ingestionFlowFiles.stream()
+                    .map(this::parseData)
+                    .flatMap(List::stream)
+                    .toList();
 
-  @Override
-  public TreasuryIufResult processFile(Long ingestionFlowFileId) {
-    List<String> iufIuvList = new ArrayList<>();
-    try {
-      IngestionFlowFileDTO ingestionFlowFileDTO = findIngestionFlowFileRecord(ingestionFlowFileId);
 
-      List<Path> ingestionFlowFiles = retrieveFiles(ingestionFlowFileDTO);
+            return new TreasuryIufResult(iufList,true);
+
+        } catch (Exception e) {
+            log.error("Error during TreasuryOpiIngestionActivity ingestionFlowFileId {}", ingestionFlowFileId, e);
+            return new TreasuryIufResult(Collections.emptyList(), false);
+        }
 
-     log.debug("Successfully retrieved the following files related to the ingestionFlowFileId {}: {}", ingestionFlowFileId, ingestionFlowFiles);
 
-    } catch (Exception e) {
-      log.error("Error during TreasuryOpiIngestionActivity ingestionFlowFileId {}", ingestionFlowFileId, e);
-      return new TreasuryIufResult(Collections.emptyList(), false);
     }
-    return new TreasuryIufResult(iufIuvList, true);
-  }
 
+    private IngestionFlowFileDTO findIngestionFlowFileRecord(Long ingestionFlowFileId) {
+        IngestionFlowFileDTO ingestionFlowFileDTO = ingestionFlowFileDao.findById(ingestionFlowFileId)
+                .orElseThrow(() -> new IngestionFlowFileNotFoundException("Cannot found ingestionFlow having id: " + ingestionFlowFileId));
+        if (!ingestionFlowFileDTO.getFlowFileType().equals(IngestionFlowFileType.OPI)) {
+            throw new IllegalArgumentException("invalid ingestionFlow file type " + ingestionFlowFileDTO.getFlowFileType());
+        }
+        return ingestionFlowFileDTO;
+    }
 
+    private List<Path> retrieveFiles(IngestionFlowFileDTO ingestionFlowFileDTO) throws IOException {
 
-  private IngestionFlowFileDTO findIngestionFlowFileRecord(Long ingestionFlowFileId) {
-      IngestionFlowFileDTO ingestionFlowFileDTO = ingestionFlowFileDao.findById(ingestionFlowFileId)
-              .orElseThrow(() -> new IngestionFlowFileNotFoundException("Cannot found ingestionFlow having id: " + ingestionFlowFileId));
-      if (!ingestionFlowFileDTO.getFlowFileType().equals(ingestionflowFileType)) {
-        throw new IllegalArgumentException("invalid ingestionFlow file type " + ingestionFlowFileDTO.getFlowFileType());
-      }
-    return ingestionFlowFileDTO;
-  }
+        return ingestionFlowFileRetrieverService
+                .retrieveAndUnzipFile(Path.of(ingestionFlowFileDTO.getFilePath()), ingestionFlowFileDTO.getFileName());
+    }
 
-  private  List<Path> retrieveFiles(IngestionFlowFileDTO ingestionFlowFileDTO) throws IOException {
+    private List<String> parseData(Path ingestionFlowFilePath) {
```

Comment:
Given the size, and probably it could grow due to the following tasks, probably it could be better if we create a new TrasuryOpiParserService

Reply:
extract methon in a new service class